Add exchange integration guide to SDK docs (#1014)
iotaledger/iota-wiki: next/docs/iota-sdk/docs/how-tos/exchange-guide.mdx

You can read more about [backup and security in this guide](https://wiki.iota.org/shimmer/introduction/how_tos/backup_security).

```bash
npm install @iota/sdk dotenv
```

:::warning

We use `dotenv` to share user-defined constants across many examples. This is for convenience only and shouldn't be done in production.

:::

You can create a `.env` by running the following command:

```bash
touch .env
```

The following examples will need these variables to exist, here are some default values you can use but also change:

```bash
WALLET_DB_PATH="./wallet-database"
NODE_URL="https://api.testnet.shimmer.network"
STRONGHOLD_SNAPSHOT_PATH="./wallet.stronghold"
STRONGHOLD_PASSWORD="*K6%79#yeFn7AMQhMQ"
EXPLORER_URL="https://explorer.shimmer.network/testnet"
```

#### 1 Generate a mnemonic

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/0-generate-mnemonic.ts
```

After you have updated the `.env` file, you can initialize the `Wallet` instance with a secret manager([`Stronghold`](https://wiki.iota.org/stronghold.rs/welcome) by default) and client options.

:::note

Manage your password with the utmost care.

:::

> By default, the Stronghold file will be called `wallet.stronghold`. It will store the seed (derived from the mnemonic) that serves as a cryptographic key from which all accounts and related addresses are generated.

Reviewer: ```suggestion
By default, the Stronghold file will be called `example.stronghold`. It will store the seed (derived from the mnemonic) that serves as a cryptographic key from which all accounts and related addresses are generated.
```

:::note

Keep the `stronghold` password and the `stronghold` database on separate devices. See the [backup and security guide](https://wiki.iota.org/introduction/how_tos/backup_security/) for more information.

:::

#### 2 Create an Account for Each User

You can import the IOTA SDK and create a wallet using the following example:

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/1-create-account.ts
```

The `Alias` must be unique and can be whatever fits your use case. The `Alias` is typically used to identify an account later on. Each account is also represented by an `index` which is incremented by one every time a new account is created. You can refer to any account via its `index`, or `alias`.

You get an instance of any created account using `Wallet.getAccount(accountId|alias)` or get all accounts with `Wallet.getAccounts()`.

Common methods of `account` instance include:

- `account.addresses()` - returns list of addresses related to the account.

> - `account.generateAddresses()` - generate a new address for the address index incremented by 1.

Reviewer: ```suggestion
- `account.generateEd25519Addresses()` - generate one or more new address(es) for the address index incremented by 1.
```

- `account.sync()` - sync the account information with the tangle.

### 3. Generate a User Address to Deposit Funds

The `wallet` module of the IOTA SDK is a stateful library. This means it caches all relevant information in storage to provide performance benefits while dealing with, potentially, many accounts and addresses.

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/2-generate-address.ts
```

Every account can have multiple addresses. Addresses are represented by an `index` which is incremented by one every time a new address is created. You can access the addresses using the `account.address()` method:

```javascript
const addresses = account.addresses();

console.log('Need a refill? Send it to this address:', addresses[0]);
```

You can use the [Faucet](https://faucet.testnet.shimmer.network/) to add test tokens and test your account.

There are two types of addresses, `internal` and `public` (external). This approach is known as a _BIP32 Hierarchical Deterministic wallet (HD Wallet)_.

- Each set of addresses is independent of each other and has an independent `index` id.

> - Addresses that are created by `account.generateAddresses()` are indicated as `internal=false` (public).

Reviewer: ```suggestion
- Addresses that are created by `account.generateEd25519Addresses()` are indicated as `internal=false` (public).
```

    metadata: [Object],
    output: [Object],
    isSpent: false,
    address: [Object],
    networkId: '1862946857608115868',
    remainder: false,
    chain: [Array]
  },
  transaction: null,
  transactionInputs: null
}
```

Alternatively you can use `account.outputs()` to get all outputs that are stored in the account, or `account.unspentOutputs()`, to get only unspent outputs.

### 6. Enable Withdrawals

You can use the following example to send tokens to an address.

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/5-send-amount.ts
```

The full function signature is `account.send(outputs[, options])`.

Default options are fine and successful; however, you can provide additional options, such as `remainderValueStrategy`, which can have the following values:

- `changeAddress`: Send the remainder value to an internal address.
- `reuseAddress`: Send the remainder value back to its original address.
- `customAddress`: Send the remainder value back to a provided account address.

```js
TransactionOptions {
    remainderValueStrategy?: RemainderValueStrategy;

>     taggedDataPayload?: ITaggedDataPayload;

Reviewer: ```suggestion
    taggedDataPayload?: TaggedDataPayload;
```

---
description: Easily integrate with your exchange, custody solution, or product using the IOTA SDK.
sidebar_label: Integrate IOTA SDK With Your Exchange
keywords:
  - integrate
  - exchange
  - account model
  - addresses
  - IOTA SDK
  - setup
  - NodeJS
  - explanation
---

import WarningUnlockConditions from '../_admonitions/_warning-unlock-conditions.md';

# Stardust Exchange Integration Guide

:::warning

The stardust update allows for complex tokenization schemes that can lead to monetary losses if misused.
Transaction outputs may have multiple unlocking conditions which may require returning some or all of the amount, which could expire if not claimed in time, or which may not be unlockable for a very long time. You can lose money if you do not check the unlock conditions before accepting a deposit!

:::

## Integration Guide

This guide explains how to integrate the IOTA SDK into your exchange, custody solution, or product.

Features of the IOTA SDK:

- Secure seed management.
- Account management with multiple accounts and multiple addresses.
- Confirmation monitoring.
- Deposit address monitoring.
- Backup and restore functionality.

### How Does it Work?

> The IOTA SDK is a stateful package with a standardized interface for developers to build value transaction applications. It offers abstractions to handle payments and can optionally interact with [Stronghold](https://wiki.iota.org/stronghold.rs/getting_started) for seed handling, seed storage, and state backup.

Reviewer 2: I don't know about the first sentence. Sounds a bit weird to me tbh. I probably would use the term `library` instead of `package`. Also I wouldn't say it's `standardized`. Do you mean `consistent`?

Reviewer 3: I don't know, didn't write this, but I can change

Reviewer 3: Changed

---
description: Easily integrate with your exchange, custody solution, or product using the IOTA SDK.
sidebar_label: Integrate IOTA SDK With Your Exchange
keywords:
  - integrate
  - exchange
  - account model
  - addresses
  - IOTA SDK
  - setup
  - NodeJS
  - explanation
---

import WarningUnlockConditions from '../_admonitions/_warning-unlock-conditions.md';

# Stardust Exchange Integration Guide

:::note

You can easily integrate the IOTA SDK in your exchange, custody solution, or product.

:::

:::warning

> The stardust update allows for complex tokenization schemes that can lead to monetary losses if misused.

Reviewer 2: ```suggestion
The Stardust update allows for complex tokenization schemes that can lead to monetary losses if misused.
```

> The stardust update allows for complex tokenization schemes that can lead to monetary losses if used incorrectly.

Author: ```suggestion
The stardust update allows for complex tokenization schemes that can lead to monetary losses if misused.
```

Transaction outputs may have multiple unlocking conditions which may require returning some or all of the amount, which could expire if not claimed in time, or which may not be unlockable for a very long time. You can lose money if you do not check the unlock conditions before accepting a deposit!

:::

## Integration Guide

> This guide explains how to integrate the IOTA SDK in your exchange.

Author: ```suggestion
This guide explains how to integrate the IOTA SDK into your exchange.
```

Features of the IOTA SDK:

- Secure seed management.
- Account management with multiple accounts and multiple addresses.
- Confirmation monitoring.
- Deposit address monitoring.
- Backup and restore functionality.

> ### How Does it Work?

Reviewer 2: ```suggestion
### How Does It Work?
```

Reviewer 3: @Dr-Electron Electron what's the policy for these? I think small connecting words are not capitalized?

Reviewer 2: I checked with this tool: https://capitalizemytitle.com/style/APA/

Author: https://wiki.iota.org/community/contribute-to-wiki/reference/style/formatting/#headings

So Alex is correct

> The IOTA SDK is a stateful package with a standardized interface for developers to build applications involving value transactions. It offers abstractions to handle payments and can optionally interact with [Stronghold](https://wiki.iota.org/stronghold.rs/getting_started) for seed handling, seed storage, and state backup.

Author: ```suggestion
The IOTA SDK is a stateful package with a standardized interface for developers to build value transaction applications. It offers abstractions to handle payments and can optionally interact with [Stronghold](https://wiki.iota.org/stronghold.rs/getting_started) for seed handling, seed storage, and state backup.
```

:::note

> If you are not familiar with the IOTA SDK, you can find more information in the [documentation](../welcome.md).

Author: ```suggestion
If you are unfamiliar with the IOTA SDK, you can find more information in the [documentation](../welcome.md).
```

:::

You can use the following examples as a guide to implementing _the multi-account approach_ using the `NodeJs` bindings:

> 1. [Set up the the IOTA SDK](#1-set-up-the-iota-sdk).

Author: ```suggestion
1. [Set up the IOTA SDK](#1-set-up-the-iota-sdk).
```

2. [Create an account for each user](#2-create-an-account-for-each-user).
3. [Generate a user address to deposit funds](#3-generate-a-user-address-to-deposit-funds).
4. [Check the user balance](#4-check-the-account-balance).
5. [Listen to events](#5-listen-to-events).
6. [Enable withdrawals](#6-enable-withdrawals).

:::note

If you are looking for other languages, please read the [IOTA SDK overview](../welcome/#your-application-in-the-iota-network).

:::

> Since all IOTA SDK bindings are based on core principles provided by the IOTA SDK, the outlined approach is very similar regardless of the programming language you choose.

Author: ```suggestion
Since all IOTA SDK bindings are based on core principles provided by the IOTA SDK, the outlined approach is very similar regardless of your chosen programming language.
```

<details>
<summary> If you were already supporting Shimmer with wallet.rs </summary>

The [IOTA SDK](https://github.com/iotaledger/iota-sdk) is a brand new project which is a combination and improvement of both [iota.rs](https://github.com/iotaledger/iota.rs) and [wallet.rs](https://github.com/iotaledger/wallet.rs).

> If you were already supporting Shimmer with wallet.rs, we encourage you to switch to the SDK as wallet.rs will be deprecated and not receive any update anymore.

Author: ```suggestion
If you were already supporting Shimmer with wallet.rs, we encourage you to switch to the SDK as wallet.rs will be deprecated and not receive any updates anymore.
```

> Most of the APIs are the same and the changes fairly minimal, here are some of the major changes:

Author: ```suggestion
Most of the APIs are the same and the changes are fairly minimal, here are some of the major changes:
```

or

```suggestion
Most of the APIs are the same and changes fairly minimal, here are some of the major changes:
```

I guess 🤔

- The npm package changes from `@iota/wallet` to `@iota/sdk`
- `AccountManager` becomes `Wallet`
- `manager.generateMnemonic()` becomes `Utils.generateMnemonic()`
- `generateAddress` becomes `generateEd25519Addresses`
- `sendAmount` becomes `send`
- The callbacks to `listen` now provide a properly typed `Event`

</details>

### 1. Set Up the IOTA SDK

> First, you should install the components that are needed to use the IOTA SDK and the binding of your choice; it may vary a bit from language to language. In the case of the [NodeJs binding](../getting-started/nodejs.mdx), it is straightforward since it is distributed via the `npm` package manager.

Author: ```suggestion
First, you should install the components needed to use the IOTA SDK and the binding of your choice; it may vary a bit from language to language. In the case of the [NodeJs binding](../getting-started/nodejs.mdx), it is straightforward since it is distributed via the `npm` package manager.
```

---
description: Easily integrate with your exchange, custody solution, or product using the IOTA SDK.
sidebar_label: Integrate IOTA SDK With Your Exchange
keywords:
  - integrate
  - exchange
  - account model
  - addresses
  - IOTA SDK
  - setup
  - NodeJS
  - explanation
---

import WarningUnlockConditions from '../_admonitions/_warning-unlock-conditions.md';

# Stardust Exchange Integration Guide

:::note

The Stardust update is a set of protocol changes that were first deployed to the Shimmer network.
See [What is Stardust?
](https://wiki.iota.org/shimmer/develop/explanations/what-is-stardust/) for an overview of its features or [TIPs](https://github.com/iotaledger/tips) flagged `Stardust` for a deeper understanding of the protocol changes.

:::

:::warning

The Stardust update allows for complex tokenization schemes that can lead to monetary losses if misused.

> Transaction outputs may have multiple unlocking conditions which may require returning some or all of the amount, which could expire if not claimed in time, or which may not be unlockable for a very long time. You can lose money if you do not check the unlock conditions before accepting a deposit!

Reviewer 4: Using multiple times `which` made me confused a little, Maybe smth  like this: 
```suggestion
Transaction outputs might have multiple unlocking conditions. These conditions may require the return of part or all of the transaction amount, have an expiration date if not claimed on time, or they may not be unlockable for an extended period of time. Therefore, you need to review these unlock conditions carefully before accepting any deposits to avoid loss of funds.
```

:::

Since all IOTA SDK bindings are based on core principles provided by the IOTA SDK, the outlined approach is very similar regardless of your chosen programming language.

<details>
<summary> If you were already supporting Shimmer with wallet.rs </summary>

The [IOTA SDK](https://github.com/iotaledger/iota-sdk) is a brand new project which is a combination and improvement of both [iota.rs](https://github.com/iotaledger/iota.rs) and [wallet.rs](https://github.com/iotaledger/wallet.rs).
If you were already supporting Shimmer with wallet.rs, we encourage you to switch to the SDK as wallet.rs will be deprecated and not receive any updates anymore.

Most of the APIs are the same and the changes are fairly minimal, here are some of the major changes:

- The npm package changes from `@iota/wallet` to `@iota/sdk`
- `AccountManager` becomes `Wallet`
- `manager.generateMnemonic()` becomes `Utils.generateMnemonic()`
- `generateAddress` becomes `generateEd25519Addresses`
- `sendAmount` becomes `send`
- The callbacks to `listen` now provide a properly typed `Event`

</details>

### 1. Set Up the IOTA SDK

> First, you should install the components needed to use the IOTA SDK and the binding of your choice; it may vary a bit from language to language. In the case of the [NodeJs binding](../getting-started/nodejs.mdx), it is straightforward since it is distributed via the `npm` package manager.

Reviewer 2: ```suggestion
First, you should install the components needed to use the IOTA SDK and the binding of your choice; it may vary a bit from language to language. In case of the [NodeJs binding](../getting-started/nodejs.mdx), it is straightforward since it is distributed via the `npm` package manager.
```

You can read more about [backup and security in this guide](https://wiki.iota.org/shimmer/introduction/how_tos/backup_security).

```bash
npm install @iota/sdk dotenv
```

:::warning

We use `dotenv` to share user-defined constants across many examples. This is for convenience only and shouldn't be done in production.

:::

You can create a `.env` by running the following command:

```bash
touch .env
```

The following examples will need these variables to exist, here are some default values you can use but also change:

```bash

> WALLET_DB_PATH="./wallet-database"

Reviewer 2: ```suggestion
WALLET_DB_PATH="./alice-database"
```

Reviewer 2: not sure whether we should propose `./example-database`, or `./alice-database` tbh. I would just use the first, but :shrug:

Reviewer 3: Nah I'd keep wallet

Reviewer 2: okay

NODE_URL="https://api.testnet.shimmer.network"

> STRONGHOLD_SNAPSHOT_PATH="./wallet.stronghold"

Reviewer 2: ```suggestion
STRONGHOLD_SNAPSHOT_PATH="./example.stronghold"
```

Reviewer 2: okay

> STRONGHOLD_PASSWORD="*K6%79#yeFn7AMQhMQ"

Reviewer 2: If find it better to use a stupid password, so it's more obvious that you should change it, but :shrug:

Reviewer 3: I don't know, I feel like stupid password is not a very good signal to send

Reviewer 2: people might think ... hey ... this a good password already. I'll keep it :trollface: 
also makes it a bit more difficult logging in using the cli-wallet, but okay.

Reviewer 2: okay

EXPLORER_URL="https://explorer.shimmer.network/testnet"
```

#### 1 Generate a mnemonic

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/0-generate-mnemonic.ts
```

After you have updated the `.env` file, you can initialize the `Wallet` instance with a secret manager([`Stronghold`](https://wiki.iota.org/stronghold.rs/welcome) by default) and client options.

:::note

Manage your password with the utmost care.

:::

By default, the Stronghold file will be called `wallet.stronghold`. It will store the seed (derived from the mnemonic) that serves as a cryptographic key from which all accounts and related addresses are generated.

> One of the key principles behind the `stronghold` is that no one can get a seed out of it, so you should also back up your 24-word mnemonic in a secure place because there is no way to recover it from the `.stronghold` file. You deal with accounts using the `Wallet` instance exclusively, where all complexities are hidden under the hood and are dealt with securely.

Reviewer 2: ```suggestion
One of the key principles behind `Stronghold` is that is impossible to extract a seed from it ever again, that is, recover it from the `.stronghold` file. That's why you should always back up your 24-word mnemonic and have it stored in safe place. You deal with accounts using the `Wallet` instance exclusively, which abstracts away all internal complexities and makes transacting in a network very easy and convenient.
```

Reviewer 3: This sentence is not factually correct. It is possible to get the seed out of stronghold.

Reviewer 3: Should probably remove that whole part

Reviewer: Replace seed with mnemonic and it's correct :trollface:

Reviewer 3: I was thinking the exact same thing ahah

:::note

Keep the `stronghold` password and the `stronghold` database on separate devices. See the [backup and security guide](https://wiki.iota.org/introduction/how_tos/backup_security/) for more information.

:::

#### 2 Create an Account for Each User

You can import the IOTA SDK and create a wallet using the following example:

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/1-create-account.ts
```

> The `Alias` must be unique and can be whatever fits your use case. The `Alias` is typically used to identify an account later on. Each account is also represented by an `index` which is incremented by one every time a new account is created. You can refer to any account via its `index`, or `alias`.

Reviewer 2: comma before `which` here as well?

> The `Alias` must be unique and can be whatever fits your use case. The `Alias` is typically used to identify an account later on. Each account is also represented by an `index` which is incremented by one every time a new account is created. You can refer to any account via its `index`, or `alias`.

Reviewer 2: ```suggestion
The `Alias` must be unique and can be whatever fits your use case. The `Alias` is typically used to identify an account later on. Each account is also represented by an `index`, which is incremented by one every time a new account is created. You can refer to any account via its `index` or its `alias`.
```

You get an instance of any created account using `Wallet.getAccount(accountId|alias)` or get all accounts with `Wallet.getAccounts()`.

> Common methods of `account` instance include:

Reviewer 2: ```suggestion
Common methods of an `account` instance include:
```

- `account.addresses()` - returns list of addresses related to the account.

> - `account.generateEd25519Addresses()` - generate one or more new address(es) for the address index incremented by 1.

Reviewer 2: ```suggestion
- `account.generateEd25519Addresses()` - generates one or more new address(es) for the address index incremented by 1.
```

- `account.balance()` - returns the balance for the given account.

> - `account.sync()` - sync the account information with the tangle.

Reviewer 2: ```suggestion
- `account.sync()` - sync the account information with the Tangle.
```

### 3. Generate a User Address to Deposit Funds

The `wallet` module of the IOTA SDK is a stateful library. This means it caches all relevant information in storage to provide performance benefits while dealing with, potentially, many accounts and addresses.

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/2-generate-address.ts
```

> Every account can have multiple addresses. Addresses are represented by an `index` which is incremented by one every time a new address is created. You can access the addresses using the `account.address()` method:

Reviewer 2: ```suggestion
Every account can have multiple addresses. Addresses are represented by an `index` which is incremented by one every time a new address is created. You can access the addresses using the `account.addresses()` method:
```

```javascript
const addresses = account.addresses();

console.log('Need a refill? Send it to this address:', addresses[0]);
```

> You can use the [Faucet](https://faucet.testnet.shimmer.network/) to add test tokens and test your account.

Reviewer 2: ```suggestion
You can use the [Faucet](https://faucet.testnet.shimmer.network/) to request test tokens and play with your account.
```

Reviewer 2: if you don't like "play with" instead of "test" ... I can edit the suggestion.

    output: [Object],
    isSpent: false,
    address: [Object],
    networkId: '1862946857608115868',
    remainder: false,
    chain: [Array]
  },
  transaction: null,
  transactionInputs: null
}
```

Alternatively you can use `account.outputs()` to get all outputs that are stored in the account, or `account.unspentOutputs()`, to get only unspent outputs.

### 6. Enable Withdrawals

You can use the following example to send tokens to an address.

```js reference
https://github.com/iotaledger/iota-sdk/tree/develop/bindings/nodejs/examples/exchange/5-send-amount.ts
```

The full function signature is `account.send(outputs[, options])`.

Default options are fine and successful; however, you can provide additional options, such as `remainderValueStrategy`, which can have the following values:

- `changeAddress`: Send the remainder value to an internal address.
- `reuseAddress`: Send the remainder value back to its original address.
- `customAddress`: Send the remainder value back to a provided account address.

```js
TransactionOptions {
    remainderValueStrategy?: RemainderValueStrategy;
    taggedDataPayload?: TaggedDataPayload;
    customInputs?: string[];
}
```

> The `account.send()` function returns a `transaction` with it's id. The `blockId` can be used later for checking a confirmation status. You can obtain individual transactions related to the given account using the `account.transactions()` function.

Reviewer 2: ```suggestion
The `account.send()` function returns a `transaction` with its id. The `blockId` can be used later for checking the confirmation status. You can obtain individual transactions related to the given account using the `account.transactions()` function.
```

<details>
<summary> If you were already supporting Shimmer with wallet.rs </summary>

The [IOTA SDK](https://github.com/iotaledger/iota-sdk) is a brand new project which is a combination and improvement of both [iota.rs](https://github.com/iotaledger/iota.rs) and [wallet.rs](https://github.com/iotaledger/wallet.rs).
If you were already supporting Shimmer with wallet.rs, we encourage you to switch to the SDK as wallet.rs will be deprecated and not receive any update anymore.

Most of the APIs are the same and the changes fairly minimal, here are some of the major changes:

- The npm package changes from `@iota/wallet` to `@iota/sdk`
- `AccountManager` becomes `Wallet`
- `manager.generateMnemonic()` becomes `Utils.generateMnemonic()`

> - `generateEd25519Address` becomes `generateEd25519Addresses`

Reviewer: ```suggestion
- `generateAddress` becomes `generateEd25519Addresses`
```

Reviewer 3: I mean, It's currently `generateEd25519Address` in the old bindings

Reviewer 3: But it's recent so maybe we should put that yes

Reviewer: Whatever you think is better, feel free to resolve
